Add mutation tests to pipeline

dkreuer opened the pull request:
Implementation for #332

dkreuer committed:
Upgrade phpunit.xml.dist

dkreuer committed:
Require roave/infection-static-analysis-plugin as dev dependency

dkreuer committed:
Add configuration for mutation testing with infection-php

dkreuer committed:
Add mutation testing pipeline

dkreuer committed:
Squash mutations

dkreuer committed:
Squash psalm errors to have perfect baseline for mutation testing

Ocramius commented:
Merging to `3.9.x`, but will rename the branch to `4.0.x` afterwards

Ocramius commented on baseline.xml:
```diff
@@ -3,17 +3,4 @@
   <file src="src/ComposerRequireChecker/Cli/CheckCommand.php">
     <MixedArgumentTypeCoercion occurrences="2"/>
```
Oh, last one left? Can we completely get rid of the baseline, perhaps?

dkreuer replied:
Currently I don’t know how to process all the ongoing generator stuff ^^

Ocramius commented on infection.json.dist:
```diff
@@ -0,0 +1,17 @@
+            "src"
+        ]
+    },
+    "timeout": 30,
```
Let's lower the timeout to 5 seconds or such: we don't have long-running integration tests anyway

dkreuer replied:
With default 10 seconds a lot mutations were skipped due to timeout. But I‘ll try to reduce it in the dist.

dkreuer replied:
Done. Many skipped tests now which reduces the MSI. They are not considered T(imed out).

Ocramius replied:
@maks-rafalko do you perhaps know if this is default behavior of @infection, or if it's a bug to be reported? :thinking:

maks-rafalko replied:
from the quick look, `S`kipped mutants should not decrease MSI:

![image](https://user-images.githubusercontent.com/3725595/145197998-0b5127c4-1d60-4364-a541-5130d29a031e.png)

but if you found a case when it does, please report

dkreuer replied:
To clarify: Skipped tests do not reduce MSI but are not accounted for in the metric at all.

I guess there are quite a few integration tests which do take time combined with possible mutation .. more than expected.

While the phpunit testsuite takes around 18-20s, mutation testing for the same test suite takes around 9m (if timeout is set to 30 and no S(kipped) mutants emerge).

Statistics with default timeout:

```
496 mutations were generated:
      57 mutants were killed
       7 mutants were not covered by tests
       2 covered mutants were not detected
       0 errors were encountered
       0 syntax errors were encountered
       0 time outs were encountered
     430 mutants required more time than configured
```

Statistics with timeout 30s:

```
496 mutations were generated:
     473 mutants were killed
       7 mutants were not covered by tests
      16 covered mutants were not detected
       0 errors were encountered
       0 syntax errors were encountered
       0 time outs were encountered
       0 mutants required more time than configured
```

Ocramius replied:
@dkreuer adding `/** @covers` to existing tests could potentially reduce the execution time by a whole lot

dkreuer replied:
Doesn't that reduce the coverage itself?

Just added a bunch of obvious `@covers` to the test classes ... result:

```
496 mutations were generated:
     384 mutants were killed
      95 mutants were not covered by tests
      17 covered mutants were not detected
       0 errors were encountered
       0 syntax errors were encountered
       0 time outs were encountered
       0 mutants required more time than configured
```

Not very promising. Probably comes from the fact that there are a lot of integration tests in contrary to unit tests. If the coverage would come from unit tests I guess `@covers` would be fine.

Ocramius replied:
Yes, but tests should be isolated to a certain unit :D 

Mutation testing integration tests is expensive, so we need to decide how coverage leaks in there.

dkreuer replied:
Possible way to move forward:
- accept the 30s timeout for now
- in subsequent issue add `@covers` And `@group integration|unit`
- infection and coverage only on `@group unit` so that coverage/MSI does not decrease

I volunteer to write additional tests ;-)

Ocramius replied:
> I volunteer to write additional tests ;-)

Cool, your problem now :D

dkreuer replied:
> Cool, your problem now :D

Not in this PR maybe please? ;-)

Ocramius commented on src/ComposerRequireChecker/FileLocator/LocateComposerPackageSourceFiles.php:
```diff
@@ -16,15 +16,22 @@
+/**
+ * @psalm-type ComposerConfig = array{vendor-dir?: string}
+ * @psalm-type ComposerRequire = array<string, string>
+ * @psalm-type ComposerAutoload = array{exclude-from-classmap?: list<string>, classmap?: list<string>, files?: list<string>, psr-0?: list<string>, psr-4?: list<string>}
```
Let's use some multi-line declarations here: everything on a single line is not healthy :D

dkreuer replied:
Done.

Ocramius commented on src/ComposerRequireChecker/JsonLoader.php:
```diff
@@ -37,6 +41,7 @@ public static function getData(string $path): array
             throw new InvalidJson('error parsing ' . $path . ': ' . $exception->getMessage(), 0, $exception);
         }
 
+        /** @psalm-var ComposerData $decodedData */
```
Is any meaningful assertion possible here? Trusting the JSON data to be valid seems a bit extreme.

I usually use @azjezz /psl for that

dkreuer replied:
I guess you mean `PSL\Json\typed()`?

i can introduce the dependency if you like.

Ocramius replied:
IMO worth adding indeed 👍 

Not sure about whether `phpstan/phpstan` will be happy about it - we'll see :D

azjezz replied:
doc: https://php-standard-library.github.io/#/components/json/

phpstan probably won't be happy :D especially if you are using Type\shape([...]), as that only works properly when using the psl psalm plugin ( https://github.com/php-standard-library/psalm-plugin/blob/main/src/EventHandler/Type/Shape/FunctionReturnTypeProvider.php )

Ocramius replied:
Yep, but dropping phpstan is no biggie either

azjezz replied:
If you would like, you can merge this PR as is, and i will send a PR similar to the ones i sent for Roave/* and Laminas/* 😉

Ocramius replied:
@dkreuer your choice on that :D

dkreuer replied:
Having help in that would be great.

azjezz replied:
cool, will send a PR for PSL integration this weekend 👍

Ocramius commented on test/ComposerRequireCheckerTest/Exception/FileParseFailedTest.php:
```diff
@@ -0,0 +1,25 @@
+use Exception;
+use PHPUnit\Framework\TestCase;
+
+final class FileParseFailedTest extends TestCase
```
Consider generally adding `/** @covers ` annotations - helps with keeping executed mutation tests small :)

dkreuer replied:
Added.

dkreuer committed:
Make pipeline fail on mutation tests

dkreuer committed:
Raise timeout for mutation tests

The timeout must be raised currently to not have integration tests (which currently count to coverage) being skipped.

dkreuer committed:
Squash mutants

Ocramius reviewed (approved):
:ship: 

Thanks @dkreuer!
